contracts/src/v0.8/dev/automation/2_1/KeeperRegistryLogicB2_1.sol

```diff
@@ -13,4 +17,224 @@ contract KeeperRegistryLogicB2_1 is KeeperRegistryBase2_1 {
     address linkNativeFeed,
     address fastGasFeed
   ) KeeperRegistryBase2_1(mode, link, linkNativeFeed, fastGasFeed) {}
+
+  ///////////////////
+  // OWNER ACTIONS //
+  ///////////////////
+
+  /**
+   * @dev Called through KeeperRegistry main contract
+   */
+  function withdrawOwnerFunds() external onlyOwner {
+    uint96 amount = s_storage.ownerLinkBalance;
+
+    s_expectedLinkBalance = s_expectedLinkBalance - amount;
+    s_storage.ownerLinkBalance = 0;
+
+    emit OwnerFundsWithdrawn(amount);
+    i_link.transfer(msg.sender, amount);
+  }
+
+  /**
+   * @dev Called through KeeperRegistry main contract
+   */
+  function recoverFunds() external onlyOwner {
+    uint256 total = i_link.balanceOf(address(this));
+    i_link.transfer(msg.sender, total - s_expectedLinkBalance);
+  }
+
+  /**
+   * @dev Called through KeeperRegistry main contract
+   */
+  function setPayees(address[] calldata payees) external onlyOwner {
+    if (s_transmittersList.length != payees.length) revert ParameterLengthError();
+    for (uint256 i = 0; i < s_transmittersList.length; i++) {
+      address transmitter = s_transmittersList[i];
+      address oldPayee = s_transmitterPayees[transmitter];
+      address newPayee = payees[i];
+      if (
+        (newPayee == ZERO_ADDRESS) || (oldPayee != ZERO_ADDRESS && oldPayee != newPayee && newPayee != IGNORE_ADDRESS)
+      ) revert InvalidPayee();
+      if (newPayee != IGNORE_ADDRESS) {
+        s_transmitterPayees[transmitter] = newPayee;
+      }
+    }
+    emit PayeesUpdated(s_transmittersList, payees);
+  }
+
+  /**
+   * @dev Called through KeeperRegistry main contract
+   */
+  function pause() external onlyOwner {
+    s_hotVars.paused = true;
+
+    emit Paused(msg.sender);
+  }
+
+  /**
+   * @dev Called through KeeperRegistry main contract
+   */
+  function unpause() external onlyOwner {
+    s_hotVars.paused = false;
+
+    emit Unpaused(msg.sender);
+  }
+
+  /**
+   * @dev Called through KeeperRegistry main contract
+   */
+  function setPeerRegistryMigrationPermission(address peer, MigrationPermission permission) external onlyOwner {
+    s_peerRegistryMigrationPermission[peer] = permission;
+  }
+
+  /////////////
+  // GETTERS //
+  /////////////
+
+  /**
+   * @notice read all of the details about an upkeep
+   */
+  function getUpkeep(uint256 id) external view returns (UpkeepInfo memory upkeepInfo) {
+    Upkeep memory reg = s_upkeep[id];
+    upkeepInfo = UpkeepInfo({
+      target: reg.target,
+      forwarder: address(reg.forwarder),
```

Comment:
same q. about interface here, it will break existing customers who use getUpkeep to query for things like balance. Alternate approach would be to expose a new function getUpkeepForwarder

Reply:
yeah good callout here. I think we try to define a consistent interface. I might prefer to do it in a follow up though.